ext4/dmverity/dmverity.go

```diff
@@ -0,0 +1,179 @@
+package dmverity
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/pkg/errors"
+
+	"github.com/Microsoft/hcsshim/ext4/internal/compactext4"
+)
+
+const (
+	blockSize = compactext4.BlockSize
+)
+
+var (
+	salt                      = bytes.Repeat([]byte{0}, 32)
+	ErrSuperBlockReadFailure  = errors.New("failed to read dm-verity super block")
+	ErrSuperBlockParseFailure = errors.New("failed to parse dm-verity super block")
+	ErrRootHashReadFailure    = errors.New("failed to read dm-verity root hash")
+)
+
+type dmveritySuperblock struct {
+	/* (0) "verity\0\0" */
+	Signature [8]byte
+	/* (8) superblock version, 1 */
+	Version uint32
+	/* (12) 0 - Chrome OS, 1 - normal */
+	HashType uint32
+	/* (16) UUID of hash device */
+	UUID [16]byte
+	/* (32) Name of the hash algorithm (e.g., sha256) */
+	Algorithm [32]byte
+	/* (64) The data block size in bytes */
+	DataBlockSize uint32
+	/* (68) The hash block size in bytes */
+	HashBlockSize uint32
+	/* (72) The number of data blocks */
+	DataBlocks uint64
+	/* (80) Size of the salt */
+	SaltSize uint16
+	/* (82) Padding */
+	_ [6]byte
+	/* (88) The salt */
+	Salt [256]byte
+	/* (344) Padding */
+	_ [168]byte
+}
+
+// VerityInfo is minimal exported version of dmveritySuperblock
+type VerityInfo struct {
+	// Offset in blocks on hash device
+	HashOffsetInBlocks int64
+	// Set to true, when dm-verity super block is also written on the hash device
+	SuperBlock    bool
+	RootDigest    string
+	Salt          string
+	Algorithm     string
+	DataBlockSize uint32
+	HashBlockSize uint32
+	DataBlocks    uint64
+}
+
+// MerkleTree constructs dm-verity hash-tree for a given byte array with a fixed salt (0-byte) and algorithm (sha256).
+func MerkleTree(data []byte) []byte {
+	layers := make([][]byte, 0)
+
+	var currentLevel = bytes.NewBuffer(data)
+
+	for currentLevel.Len() != blockSize {
+		var blocks = currentLevel.Len() / blockSize
+		var nextLevel = bytes.NewBuffer(make([]byte, 0))
+
+		for i := 0; i < blocks; i++ {
+			block := make([]byte, blockSize)
+			_, _ = currentLevel.Read(block)
+			h := hash2(salt, block)
+			nextLevel.Write(h)
+		}
+
+		padding := bytes.Repeat([]byte{0}, blockSize-(nextLevel.Len()%blockSize))
+		nextLevel.Write(padding)
+
+		currentLevel = nextLevel
+		layers = append(layers, currentLevel.Bytes())
```

Comment:
Why don't we write the first layer?

Reply:
the first layer is the VHD data itself, so we're skipping it here.

Reply:
what do you mean by "first layer" @katiewasnothere?